scripts/dp_stats/dp_svm.py

```diff
@@ -1,120 +1,151 @@
+from .general_funcs import noisevector
 
-def noisevector( scale, Length ):
 
-    r1 = np.random.normal(0, 1, Length)#standard normal distribution
-    n1 = np.linalg.norm( r1, 2 )#get the norm of this random vector
-    r2 = r1 / n1#the norm of r2 is 1
-    normn = np.random.gamma( Length, 1/scale, 1 )#Generate the norm of noise according to gamma distribution
-    res = r2 * normn#get the result noise vector
-    return res
-
-def huber(z, h):#chaudhuri2011differentially corollary 21
-
-    if z > 1 + h:
-        hb = 0
-    elif np.fabs(1-z) <= h:
-        hb = (1 + h - z)**2 / (4 * h)
+def huberloss(z, huberconst):  
+    # chaudhuri2011differentially: equation 7 & corollary 13
+    if z > 1.0 + huberconst:
+        hz = 0
+    elif z < 1.0 - huberconst:
+        hz = 1 - z
     else:
-        hb = 1 - z
-    return hb
-
-def svm_output_train(data, labels, epsilon, Lambda, h):
-
-    N = len( labels )
-    l = len( data[0] )#length of a data point
-    scale = N * Lambda * epsilon / 2
-    noise = noisevector( scale, l )
-    x0 = np.zeros(l)#starting point with same length as any data point
-
-    def obj_func(x):
-        jfd = huber( labels[0] * np.dot(data[0],x), h )
-        for i in range(1,N):
-            jfd = jfd + huber( labels[i] * np.dot(data[i],x), h )
-        f = ( 1/N )*jfd + (1/2) * Lambda * ( np.linalg.norm(x,2)**2 )
-        return f
-
-    #minimization procedure
-    f = minimize(obj_func, x0, method='Nelder-Mead').x #empirical risk minimization using scipy.optimize minimize function
-    fpriv = f + noise
-    return fpriv
-
-def svm_objective_train(data, labels,  epsilon, Lambda, h):
-
-    #parameters in objective perturbation method
-    c = 1 / ( 2 * h )#chaudhuri2011differentially corollary 13
-    N = len( labels )#number of data points in the data set
-    l = len( data[0] )#length of a data point
-    x0 = np.zeros(l)#starting point with same length as any data point
-    Epsilonp = epsilon - 2 * np.log( 1 + c / (Lambda * N))
-    if Epsilonp > 0:
-        Delta = 0
-    else:
-        Delta = c / ( N * (np.exp(epsilon/4)-1) ) - Lambda
-        Epsilonp = epsilon / 2
-    noise = noisevector(Epsilonp/2, l)
-
-    def obj_func(x):
-        jfd = huber( labels[0] * np.dot(data[0], x), h)
-        for i in range(1,N):
-            jfd = jfd + huber( labels[i] * np.dot(data[i], x), h )
-        f = (1/N) * jfd + (1/2) * Lambda * (np.linalg.norm(x,2)**2) + (1/N) * np.dot(noise,x) + (1/2)*Delta*(np.linalg.norm(x,2)**2)
-        return f
-
-    #minimization procedure
-    fpriv = minimize(obj_func, x0, method='Nelder-Mead').x#empirical risk minimization using scipy.optimize minimize function
-    return fpriv
-
-def dp_svm(data, labels, method='obj', epsilon=0.1, Lambda = 0.01, h = 0.5):
+        hz = (1 + huberconst - z)**2 / (4 * huberconst)
+    return hz
+
+def eval_svm(weights, XY, num, lambda_, b, huberconst):
+    # Return the svm loss.
+
+    # add huber loss from all samples
+    XY = np.matmul(XY, weights)
```

Comment:
Better to set this to a new variable rather than overwriting `XY`.

In Python, parameters to a function are passed in by reference. That means that if you modify a parameter inside a function, it's modified in the outer scope as well, which can have unintended consequences. It's also easier to understand what's happening in the code if you assign the result to a new variable name.